File: dart/lib/src/logger.dart
/// Logger implementing the silent success pattern.
import 'dart:io';

class Logger {
  final bool _isVerbose;
  final List<String> _buffer = [];

  Logger([bool? verbose])
      : _isVerbose = verbose ??
            Platform.environment['VERBOSE'] == 'true' ||
                Platform.executableArguments.contains('--verbose') ||
                Platform.executableArguments.contains('-v');

  bool get isVerbose => _isVerbose;

  void log(Object? message) {
    if (_isVerbose) {
      print(message);
    } else {
      _buffer.add(message?.toString() ?? 'null');
    }
  }

  void warn(Object? message) {
    if (_isVerbose) {
      print('\x1B[33m$message\x1B[0m');
    } else {
      _buffer.add('\x1B[33m$message\x1B[0m');
    }
  }

  void error(Object? message) {
    print('\x1B[31m$message\x1B[0m');
  }

  void dump() {
    if (!_isVerbose && _buffer.isNotEmpty) {
      for (final msg in _buffer) {
        print(msg);
      }
      _buffer.clear();
    }
  }
}

final logger = Logger();

File: ts/src/logger.ts
/**
 * Candi Script Logger
 *
 * Implements "Silent Success, Roaring Failures" pattern.
 * Buffers logs and only outputs them if an error occurs or VERBOSE=true.
 */

export interface Logger {
    log(...args: unknown[]): void;
    error(...args: unknown[]): void;
    warn(...args: unknown[]): void;
    dump(): void;
    readonly isVerbose: boolean;
}

const isVerbose =
    process.env['VERBOSE'] === 'true' ||
    process.argv.includes('--verbose') ||
    process.argv.includes('-v');

const buffer: string[] = [];

export const logger: Logger = {
    log: (...args: unknown[]) => {
        if (isVerbose) {
            console.log(...args);
        } else {
            buffer.push(args.map(String).join(' '));
        }
    },
    error: (...args: unknown[]) => {
        // Errors are always "roaring"
        console.error(...args);
    },
    warn: (...args: unknown[]) => {
        if (isVerbose) {
            console.warn(...args);
        } else {
            buffer.push(`[WARN] ${args.map(String).join(' ')}`);
        }
    },
    dump: () => {
        if (!isVerbose && buffer.length > 0) {
            console.log('\n--- Captured Logs ---');
            buffer.forEach((line) => console.log(line));
            console.log('--- End of Logs ---\n');
        }
    },
    get isVerbose() {
        return isVerbose;
    },
};
